client/iOS/FreeRDP/ios_cliprdr.m

```diff
@@ -0,0 +1,503 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Android Clipboard Redirection
+ *
+ * Copyright 2013 Felix Long
+ * Copyright 2015 Thincast Technologies GmbH
+ * Copyright 2015 DI (FH) Martin Haimberger <martin.haimberger@thincast.com>
+ * Copyright 2023 Iordan Iordanov
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <freerdp/config.h>
+
+#include <winpr/crt.h>
+#include <winpr/stream.h>
+
+#include <freerdp/client/channels.h>
+#include <freerdp/client/cliprdr.h>
+
+#include "ios_cliprdr.h"
+
+UINT ios_cliprdr_send_client_format_list(CliprdrClientContext* cliprdr)
+{
+	UINT rc = ERROR_INTERNAL_ERROR;
+	UINT32 index;
+	UINT32 formatId;
+	UINT32 numFormats;
+	UINT32* pFormatIds;
+	const char* formatName;
+	CLIPRDR_FORMAT* formats;
+	CLIPRDR_FORMAT_LIST formatList = { 0 };
+
+	if (!cliprdr)
+		return ERROR_INVALID_PARAMETER;
+
+	mfContext* afc = (mfContext*)cliprdr->custom;
+
+	if (!afc || !afc->cliprdr)
+		return ERROR_INVALID_PARAMETER;
+
+	pFormatIds = NULL;
+	numFormats = ClipboardGetFormatIds(afc->clipboard, &pFormatIds);
+	formats = (CLIPRDR_FORMAT*)calloc(numFormats, sizeof(CLIPRDR_FORMAT));
+
+	if (!formats)
+		goto fail;
+
+	for (index = 0; index < numFormats; index++)
+	{
+		formatId = pFormatIds[index];
+		formatName = ClipboardGetFormatName(afc->clipboard, formatId);
+		formats[index].formatId = formatId;
+		formats[index].formatName = NULL;
+
+		if ((formatId > CF_MAX) && formatName)
+		{
+			formats[index].formatName = _strdup(formatName);
+
+			if (!formats[index].formatName)
+				goto fail;
+		}
+	}
+
+	formatList.common.msgFlags = CB_RESPONSE_OK;
```

Comment:
This flag is wrong and does not conform to `[MS-RDPECLIP]`. Only allowed flag is the assci_names flag.

Reply:
Same as the previous response.

Reply:
No, that's not how it works. You introduce a protocol violation here.

Reply:
@pnowack formatList.common.msgFlags = CB_RESPONSE_OK; is exactly what is called in client/Android/Studio/freeRDPCore/src/main/cpp/android_cliprdr.c.

What shall we do about android_cliprdr?

Are you referring to:

#define CB_ASCII_NAMES 0x0004

Shall we change both files then?

Reply:
that line can be removed, the `cliprdr_packet_format_list_new` ignores this field anyway.
the `CB_ASCII_NAMES` state from capability exchange is stored internally and used to determine if short or long names can be used.